### PR TITLE
shell: add ddd_diagnose command

### PR DESCRIPTION
**Pull request by qinzuoyan**
refer to https://github.com/XiaoMi/rdsn/issues/80

### CODE REVIEW

File: src/shell/commands.h [hunk 1]
```diff
@@ -4310,3 +4310,228 @@ inline bool clear_app_envs(command_executor *e, shell_context *sc, arguments arg
...
+    if (ret != dsn::ERR_OK) {
+        fprintf(stderr, "ERROR: DDD diagnose failed with err = %s\n", ret.to_string());
+        return true;
+    }
+
+    std::streambuf *buf;
+    std::ofstream of;
+
+    if (!out_file.empty()) {
+        of.open(out_file);
```
Comment (acelyc111) [inline src/shell/commands.h]: of使用完毕需要close掉

File: src/shell/commands.h [hunk 2]
```diff
@@ -4310,3 +4310,228 @@ inline bool clear_app_envs(command_executor *e, shell_context *sc, arguments arg
...
+            }
+        }
+        out << std::endl;
+        out << "Proposed count: " << proposed_count << "/" << ddd_partitions.size() << std::endl;
+        out << std::endl;
+    }
+
+    std::cout << "Diagnose ddd done." << std::endl;
+    return true;
+}
```
Comment (neverchanje) [inline src/shell/commands.h]: add newline

Comment (qinzuoyan) [reply src/shell/commands.h]: fixed

File: src/shell/commands.h (hunk 1 above)
Comment (qinzuoyan) [reply src/shell/commands.h]: 不用自己close，对象析构时会自动close

File: src/shell/commands.h [hunk 3]
```diff
@@ -4310,3 +4310,228 @@ inline bool clear_app_envs(command_executor *e, shell_context *sc, arguments arg
     }
     return true;
 }
+
+inline dsn::rpc_address diagnose_recommend(const ddd_partition_info &pinfo)
```
[the remark this answers is not in the document]
Comment (qinzuoyan) [reply src/shell/commands.h]: fixed

**Pull request merged**